### PR TITLE
added ComputeAABB func

### PR DESCRIPTION
**Pull request by soap1984**

### CODE REVIEW

File: src/src/SceneNode.cpp [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
 #include <mineola/Renderable.h>
+#include <glm/gtc/matrix_transform.hpp>
+
+namespace {
+float boolToSign(bool b) {
+  return b ? 1.f : -1.f;
+}
+
+// index range 0~7
+void AABBVertexes(const glm::vec3 &min, const glm::vec3 &max, std::vector<glm::vec3> &vts) {
```
Comment (shenfy) [inline src/src/SceneNode.cpp:15]: make this a member function of the AABB class, vector<vec3> AABB::Corners()
add comments to make it more readable.

File: src/src/SceneNode.cpp [hunk 2]
```diff
@@ -1,8 +1,28 @@
...
+  auto center = (min + max) * 0.5f;
+  auto extent = max - min;
+
+  vts.clear();
+  for (int i = 0; i <= 7; i++) {
```
Comment (shenfy) [inline src/src/SceneNode.cpp:20]: use i < 8, so it's more readable to human (8 vertices in a cube)

File: src/src/SceneNode.cpp [hunk 3]
```diff
@@ -1,8 +1,28 @@
 #include "prefix.h"
+#include <vector>
 #include <mineola/Light.h>
 #include <mineola/Camera.h>
 #include <mineola/SceneNode.h>
 #include <mineola/Renderable.h>
+#include <glm/gtc/matrix_transform.hpp>
+
+namespace {
+float boolToSign(bool b) {
```
Comment (shenfy) [inline src/src/SceneNode.cpp:10]: Bool2Sign

File: src/src/SceneNode.cpp [hunk 4]
```diff
@@ -196,4 +216,36 @@ namespace mineola {
...
+
+void SceneNode::ComputeAABB(glm::vec3 &center, glm::vec3 &extent) {
+  glm::vec3 min(std::numeric_limits<float>::max()), max(std::numeric_limits<float>::lowest());
+
+  DFTraverse([&min, &max](SceneNode &node) {
+    bool valid_aabb = false;
+    glm::vec3 min_node(std::numeric_limits<float>::max()), max_node(std::numeric_limits<float>::lowest());
+    for (auto renderable : node.Renderables()) {
+      // check min and max vertex is valid or not
+      if (renderable->MinVertex() != glm::vec3() || renderable->MaxVertex() != glm::vec3()) {
```
Comment (shenfy) [inline src/src/SceneNode.cpp:228]: check is renderable has valid AABB (i.e. unique_ptr is not null). glm::vec3() is a valid vertex value, shouldn't be used as a flag.

File: src/src/SceneNode.cpp [hunk 5]
```diff
@@ -196,4 +216,36 @@ namespace mineola {
         return node.Name() == name_str;
       });
   }
+
+void SceneNode::ComputeAABB(glm::vec3 &center, glm::vec3 &extent) {
+  glm::vec3 min(std::numeric_limits<float>::max()), max(std::numeric_limits<float>::lowest());
+
+  DFTraverse([&min, &max](SceneNode &node) {
+    bool valid_aabb = false;
+    glm::vec3 min_node(std::numeric_limits<float>::max()), max_node(std::numeric_limits<float>::lowest());
```
Comment (shenfy) [inline src/src/SceneNode.cpp:225]: add a AABB::Combine(const AABB &other) method instead

File: src/src/SceneNode.cpp [hunk 6]
```diff
@@ -1,8 +1,28 @@
...
+
+// index range 0~7
+void AABBVertexes(const glm::vec3 &min, const glm::vec3 &max, std::vector<glm::vec3> &vts) {
+  auto center = (min + max) * 0.5f;
+  auto extent = max - min;
+
+  vts.clear();
+  for (int i = 0; i <= 7; i++) {
+    glm::vec3 sign(boolToSign(i & 1), boolToSign(i & 2), boolToSign(i & 4));
+    vts.push_back(center + extent * 0.5f * sign);
```
Comment (shenfy) [inline src/src/SceneNode.cpp:22]: use return value of vector<vec3> instead of out parameter.

**Review by shenfy — CHANGES_REQUESTED** [review]

File: src/include/mineola/Renderable.h [hunk 7]
```diff
@@ -10,6 +10,16 @@
 
 namespace mineola {
 
+struct AABB {
+  glm::vec3 lb_{0.f};
```
Comment (shenfy) [inline src/include/mineola/Renderable.h:14]: for struct public members, do not use the trailing underscore.

File: src/include/mineola/Renderable.h [hunk 8]
```diff
@@ -36,6 +46,9 @@ class Renderable : public Resource {
 
   void SetSkin(std::shared_ptr<Skin> skin);
 
+  void SetAABB(const glm::vec3 &lb, const glm::vec3 &ub);
```
Comment (shenfy) [inline src/include/mineola/Renderable.h:49]: SetAABB should take an AABB as argument

File: src/src/Renderable.cpp [hunk 9]
```diff
@@ -5,8 +5,55 @@
...
+}
+
+glm::vec3 AABB::Extent() const {
+  return ub_ - lb_;
+}
+
+std::vector<glm::vec3> AABB::Corners() const {
+  auto center = Center();
+  auto extent = Extent();
+  auto corners = std::vector<glm::vec3>();
```
Comment (shenfy) [inline src/src/Renderable.cpp:27]: for fixed length array, set its length at construction time.

File: src/src/Renderable.cpp [hunk 10]
```diff
@@ -65,6 +112,16 @@ const std::string &Renderable::GetMaterialName(size_t index) const {
   return material_names_[index];
 }
 
+void Renderable::SetAABB(const glm::vec3 &lb, const glm::vec3 &ub) {
+  bbox_ = std::make_unique<AABB>();
+  bbox_->lb_ = lb;
+  bbox_->ub_ = ub;
+}
+
+const std::unique_ptr<AABB> &Renderable::GetAABB() const {
```
Comment (shenfy) [inline src/src/Renderable.cpp:121]: name it AABB instead of GetAABB

File: src/src/SceneNode.cpp [hunk 11]
```diff
@@ -196,4 +198,35 @@ namespace mineola {
...
+  AABB bbox;
+  bbox.lb_ = glm::vec3(std::numeric_limits<float>::max());
+  bbox.ub_ = glm::vec3(std::numeric_limits<float>::lowest());
+
+  DFTraverse([&bbox](SceneNode &node) {
+    bool valid_aabb = false;
+    AABB bbox_node;
+    bbox_node.lb_ = glm::vec3(std::numeric_limits<float>::max());
+    bbox_node.ub_ = glm::vec3(std::numeric_limits<float>::lowest());
+    glm::vec3 min_node(std::numeric_limits<float>::max()), max_node(std::numeric_limits<float>::lowest());
```
Comment (shenfy) [inline src/src/SceneNode.cpp:212]: remove unused variables

File: src/src/SceneNode.cpp [hunk 12]
```diff
@@ -196,4 +198,35 @@ namespace mineola {
...
+
+void SceneNode::ComputeAABB(glm::vec3 &center, glm::vec3 &extent) {
+  AABB bbox;
+  bbox.lb_ = glm::vec3(std::numeric_limits<float>::max());
+  bbox.ub_ = glm::vec3(std::numeric_limits<float>::lowest());
+
+  DFTraverse([&bbox](SceneNode &node) {
+    bool valid_aabb = false;
+    AABB bbox_node;
+    bbox_node.lb_ = glm::vec3(std::numeric_limits<float>::max());
```
Comment (shenfy) [inline src/src/SceneNode.cpp:210]: use the AABB(lb, ub) constructor

File: src/src/SceneNode.cpp [hunk 13]
```diff
@@ -196,4 +198,35 @@ namespace mineola {
         return node.Name() == name_str;
       });
   }
+
+void SceneNode::ComputeAABB(glm::vec3 &center, glm::vec3 &extent) {
```
Comment (shenfy) [inline src/src/SceneNode.cpp:202]: return an AABB instead.

**Review by shenfy — CHANGES_REQUESTED** [review]

**Review by chubei — CHANGES_REQUESTED** [review]

File: src/src/SceneNode.cpp [hunk 14]
```diff
@@ -196,4 +198,31 @@ namespace mineola {
...
+
+    if (valid_aabb) {
+      auto model_mat = glm::scale(node.Rbt().ToMatrix(), node.Scale());
+      bbox_node.Transform(model_mat);
+      bbox.Combine(bbox_node);
+    }
+  });
+
+  return bbox;
+}
```
Comment (chubei) [inline src/src/SceneNode.cpp:226]: std::optional<AABB> result;
for (auto &child: children) {
  auto aabb = child.ComputeAABB();
  if (aabb) {
    if (result) {
      result.Combine(aabb);
    } else {
      result = aabb;
    }
  }
}

if (result) {
    // apply model mat
}

return result;

File: src/src/Renderable.cpp [hunk 15]
```diff
@@ -1,12 +1,10 @@
 #include "prefix.h"
 #include <mineola/Renderable.h>
-#include <glm/gtc/type_ptr.hpp>
 #include <mineola/Engine.h>
 #include <mineola/Material.h>
 #include <mineola/VertexType.h>
 
 namespace mineola {
-
```
Comment (chubei-oppen) [inline src/src/Renderable.cpp:9]: don't remove this blank line

File: src/include/mineola/Renderable.h [hunk 16]
```diff
@@ -36,6 +38,9 @@ class Renderable : public Resource {
 
   void SetSkin(std::shared_ptr<Skin> skin);
 
+  void SetBbox(const AABB &bbox);
+  std::optional<AABB> Bbox() const;
```
Comment (chubei-oppen) [inline src/include/mineola/Renderable.h:42]: const std::optional<AABB> &

File: src/src/SceneNode.cpp [hunk 17]
```diff
@@ -196,4 +198,39 @@ namespace mineola {
         return node.Name() == name_str;
       });
   }
+
+std::optional<AABB> SceneNode::ComputeAABB() const {
+  AABB bbox{glm::vec3(std::numeric_limits<float>::max()),
+    glm::vec3(std::numeric_limits<float>::lowest())};
+
```
Comment (chubei-oppen) [inline src/src/SceneNode.cpp:205]: remove dead code

**Review by chubei-oppen — CHANGES_REQUESTED** [review]

**Review by chubei-oppen — APPROVED** [review]

**Pull request merged**